### PR TITLE
HFX-1715: Backport of the TLS 1.2 support to XS 6.5 SP1

### PR DESCRIPTION
**Pull request by fillzero**
This is for TLS1.2 HFX on LCM branches.

These commits originated on trunk and then was backported to cream-lcm branch, now cherry-picked from cream-lcm to cream-bugfix branch.

@thomassa , would you kindly review this?

### CODE REVIEW

File: ocaml/idl/datamodel.ml [hunk 1]
```diff
@@ -69,13 +68,18 @@ let cream_release_schema_minor_vsn = 73
 let indigo_release_schema_major_vsn = 5
 let indigo_release_schema_minor_vsn = 74
 
-let pool_health_check_release_schema_major_vsn = 5
-let pool_health_check_release_schema_minor_vsn = 75
+let cream_tls_1_2_release_schema_major_vsn = 5
+let cream_tls_1_2_release_schema_minor_vsn = 75
```
Comment (thomassa) [inline ocaml/idl/datamodel.ml:72]: We should not be re-using schema version numbers like this: we should keep the `pool_health_check` schema version, and add a `cream_tls_1_2_release` schema version with a larger number.

File: ocaml/idl/datamodel_types.ml [hunk 2]
```diff
@@ -52,7 +52,7 @@ let rel_clearwater_whetstone = "clearwater-whetstone"
 let rel_creedence = "creedence"
 let rel_cream = "cream"
 let rel_indigo = "indigo"
-let rel_pool_health_check = "pool-health-check"
+let rel_hfx_tls1_2 = "hfx_tls1_2"
```
Comment (thomassa) [inline ocaml/idl/datamodel_types.ml:55]: Why have you removed `rel_pool_health_check` instead of just adding `rel_hfx_tls1_2` as an extra item?

File: ocaml/idl/datamodel_types.ml [hunk 3]
```diff
@@ -73,7 +73,7 @@ let release_order =
 	; rel_creedence
 	; rel_cream
 	; rel_indigo
-	; rel_pool_health_check
+	; rel_hfx_tls1_2
```
Comment (thomassa) [inline ocaml/idl/datamodel_types.ml:76]: I think we should keep `rel_pool_health_check` and add `rel_hfx_tls1_2` to the end of the list.
(Or are we still waiting to release something Cream-based with the health-check feature?)

File: ocaml/xapi/xapi.ml [hunk 4]
```diff
@@ -708,6 +708,34 @@ let set_stunnel_timeout () =
   with _ ->
     debug "Using default stunnel timeout (usually 43200)"
 
+(* Consult inventory, because to do DB lookups we must contact the
+ * master, and to do that we need to start an outgoing stunnel. *)
+let set_stunnel_legacy_inv ~__context () =
+  Stunnel.set_good_ciphersuites (match !Xapi_globs.ciphersuites_good_outbound with
+    | "" -> raise (Api_errors.Server_error (Api_errors.internal_error,["Configuration file does not specify ciphersuites-good-outbound."]))
```
Comment (thomassa) [inline ocaml/xapi/xapi.ml:715]: If the file does not specify `ciphersuites-good-outbound` at all then we use the built-in non-blank default, and we do not raise this error.
The error string here should be "Configuration file specifies a blank string for ciphersuites-good-outbound."

File: ocaml/xapi/xapi_config.ml [hunk 5]
```diff
@@ -37,7 +37,13 @@ let read_config filename =
 		"igd-passthru-vendor-whitelist", Config.Set_string vendor_whitelist;
 		"relax-xsm-sr-check", Config.Set_bool Xapi_globs.relax_xsm_sr_check;
 		"hotfix-fingerprint", Config.Set_string Xapi_globs.trusted_patch_key;
-		"gpumon_stop_timeout", Config.Set_float Xapi_globs.gpumon_stop_timeout;
+		(let cgo = "ciphersuites-good-outbound" in
```
Comment (thomassa) [inline ocaml/xapi/xapi_config.ml:40]: You have removed the entry for `gpumon_stop_timeout`. This looks like a mistake.

File: ocaml/xapi/xapi_globs.ml [hunk 6]
```diff
@@ -730,7 +731,7 @@ let redo_log_max_startup_time = ref 5.
 (** The delay between each attempt to connect to the block device I/O process *)
 let redo_log_connect_delay = ref 0.1
 
-let gpumon_stop_timeout = ref 10.0
+let gpumon_stop_timeout = ref 60.0
```
Comment (thomassa) [inline ocaml/xapi/xapi_globs.ml:734]: Why is this change part of this pull-request?

File: ocaml/xapi/xapi_globs.ml [hunk 7]
```diff
@@ -14,6 +14,7 @@
 
 (** A central location for settings related to xapi *)
 
+module String_plain = String (* For when we don't want the Xstringext version *)
```
Comment (thomassa) [inline ocaml/xapi/xapi_globs.ml:17]: Do we use this? Is it needed?

File: ocaml/xapi/xapi_gpumon.mli [hunk 8]
```diff
@@ -14,4 +14,4 @@
 
 (** Stop gpumon if it's running, perform f, then start gpumon if
   * no other threads which require gpumon to be stopped are running. *)
-val with_gpumon_stopped : ?timeout:float -> (unit -> 'a) -> 'a
+val with_gpumon_stopped : (unit -> 'a) -> 'a
```
Comment (thomassa) [inline ocaml/xapi/xapi_gpumon.mli:17]: Why is this change in this pull-request? It appears unrelated.

**Review by thomassa — CHANGES_REQUESTED** [review]
I found a few problems.
Are any of these on any other branches as well?

File: ocaml/xapi/xapi_config.ml (hunk 5 above)
Comment (fillzero) [reply ocaml/xapi/xapi_config.ml:40]: Hi @thomassa ,
This change is from below commit:

commit 699ae79d6a1169754abaf4eee1090143dc491f24
Author: Thomas Sanders thomas.sanders@citrix.com
Date:   Fri Mar 4 16:06:50 2016 +0000

```
CP-14701: Handle ciphersuite config in xapi_config

Fix a build-blocker by adapting a code-change backported from
the upstream xapi-project/xen-api git repository, which handles
the ciphersuite config in xapi_globs.ml instead.

Signed-off-by: Thomas Sanders <thomas.sanders@citrix.com>
```

diff --git a/ocaml/xapi/xapi_config.ml b/ocaml/xapi/xapi_config.ml
index 9a0fd46..e6f9a8d 100644
--- a/ocaml/xapi/xapi_config.ml
+++ b/ocaml/xapi/xapi_config.ml

```
@@ -37,7 +37,13 @@ let read_config filename =
                "igd-passthru-vendor-whitelist", Config.Set_string vendor_whitelist;
                "relax-xsm-sr-check", Config.Set_bool Xapi_globs.relax_xsm_sr_check;
                "hotfix-fingerprint", Config.Set_string Xapi_globs.trusted_patch_key;
-               "gpumon_stop_timeout", Config.Set_float Xapi_globs.gpumon_stop_timeout;
+               (let cgo = "ciphersuites-good-outbound" in
+               cgo, Config.String (fun s ->
+                       D.debug "Processing config %s=%s" cgo s;
+                       Xapi_globs.ciphersuites_good_outbound :=
+                               (if String.filter_chars s (not ++ String.isspace) <> "" then Some s else None)
+               ));
+               "ciphersuites-legacy-outbound", Config.Set_string Xapi_globs.ciphersuites_legacy_outbound;
```

File: ocaml/xapi/xapi_globs.ml (hunk 7 above)
Comment (fillzero) [reply ocaml/xapi/xapi_globs.ml:17]: @thomassa ,
This change is from below commit:

commit 12608d8aa3f6e68355b5b8a2fadf781887f4f9a0
Author: Thomas Sanders thomas.sanders@citrix.com
Date:   Fri Jan 22 17:01:00 2016 +0000

```
CP-14531: Configurable outgoing TLS ciphersuites

Instead of being hard-coded in the Stunnel library module, these are
read from xapi.conf and passed to the Stunnel module on xapi start-up.

Signed-off-by: Thomas Sanders <thomas.sanders@citrix.com>
(cherry picked from commit b492b389318743a52b2129923172a3daf43a89b6)

Conflicts:

    ocaml/xapi/xapi_globs.ml
    scripts/xapi.conf

Signed-off-by: Thomas Sanders <thomas.sanders@citrix.com>
```

```
diff --git a/ocaml/xapi/xapi_globs.ml b/ocaml/xapi/xapi_globs.ml
index e63fd3b..baba3c6 100644
--- a/ocaml/xapi/xapi_globs.ml
+++ b/ocaml/xapi/xapi_globs.ml
@@ -14,6 +14,7 @@

 (** A central location for settings related to xapi *)

+module String_plain = String (* For when we don't want the Xstringext version *)
 open Stringext
 open Printf
```

File: ocaml/xapi/xapi_gpumon.mli (hunk 8 above)
Comment (fillzero) [reply ocaml/xapi/xapi_gpumon.mli:17]: Commit "CA-209508: Use timeout when attempting to stop gpumon " was already in cream-bugfix branch.  I deleted this in PR #195 ,  please review #195.

File: ocaml/xapi/xapi_globs.ml (hunk 6 above)
Comment (fillzero) [reply ocaml/xapi/xapi_globs.ml:734]: Commit "CA-209508: Use timeout when attempting to stop gpumon" was already in cream-bugfix branch. I deleted this in PR #195 , please review #195.

File: ocaml/idl/datamodel.ml (hunk 1 above)
Comment (fillzero) [reply ocaml/idl/datamodel.ml:72]: This was introduced by manual merge conflict  "CP-14804: Add an indigo release to the datamodel", Fixed in #195

File: ocaml/idl/datamodel_types.ml (hunk 2 above)
Comment (fillzero) [reply ocaml/idl/datamodel_types.ml:55]: This was introduced by manual merge conflict "CP-14804: Add an indigo release to the datamodel", Fixed in #195

File: ocaml/idl/datamodel_types.ml (hunk 3 above)
Comment (fillzero) [reply ocaml/idl/datamodel_types.ml:76]: Fixed this in #195

File: ocaml/xapi/xapi_globs.ml (hunk 7 above)
Comment (thomassa) [reply ocaml/xapi/xapi_globs.ml:17]: That commit also adds code that uses the module.

But I've found a later commit 8dff3e9c88d0017fb50227ee33714d77e26e67d8 in which I moved that code to another file, thus rendering `String_plain` obsolete. My mistake, but let's leave it now that it's deep among other things in the revision history.

**Comment by fillzero** [comment]
Hi @thomassa  ,
Almost of these problems were introduced by conflicts when rebase to bugfix branch, some commits were already in bugfix branch, but still in rebase list due to file conflict.  And there was no conflict when rebase after I excluded these already merged commits. Please review #195 .

File: ocaml/xapi/xapi_config.ml (hunk 5 above)
Comment (thomassa) [reply ocaml/xapi/xapi_config.ml:40]: No, we definitely need to keep the `gpu_stop_timeout` line.
My original commit only added the ciphersuites-related lines here, but when you applied my change to a different base you accidentally removed the `gpu_stop_timeout` line.

The commit 699ae79d6a1169754abaf4eee1090143dc491f24 that you quote above has me as "author" but was committed by you, as a change to code that included the `gpumon_stop_timeout` line.

My original commit added some lines in that chunk but did not remove `gpumon_stop_timeout` because it was not there to start with. This was commit 25a99e89f047abbd6d4a4a576c4f2d0ce4302628 on my own branch: see https://github.com/xenserver/xen-api/pull/68/commits/25a99e89f047abbd6d4a4a576c4f2d0ce4302628

With the `gpu_stop_timeout` line missing, the config file cannot be used to set this timeout.

**Comment by thomassa** [comment]
Closing this pullreq because it is superseded by #195.
